peps/pep-0705.rst

```diff
@@ -185,111 +381,180 @@ Discussion:
         x: int
+        y: str
+
+    a: A = { "x": 1, "y": "foo" }
+    a.popitem()  # Accepted by type checker
+    a.clear()  # Accepted by type checker
+
+``update`` has been difficult to type correctly due to the open nature of TypedDict objects. Keys not specified on the type could still be present (and constrained) due to structural subtyping, meaning type safety could be accidentally violated. For instance::
 
-    class B(TypedMapping):
+    class B(TypedDict, total=False):
         x: int
+
+    def update_b(b1: B, b2: B) -> None:
+        b1.update(b2)
+
+    class C(B, TypedDict, total=False):
+        y: int
+
+    class D(B, TypedDict, total=False):
         y: str
 
-    def sum_values(m: Mapping[str, int]) -> int:
-        return sum(m.values())
+    c: C = { "x": 1, "y": 2 }
+    d: D = { "x": 3, "y": "foo" }
+    update_b(c, d)  # c is no longer a C at runtime
 
-    def f(a: A) -> None:
-        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]
+Both mypy and pyright currectly permit this usage, however, as the only viable alternative has been to prevent calling ``update`` at all.
 
-    b: B = {'x': 0, 'y': 'foo'}
-    f(b)  # Runtime error: int + str
+With the addition of ``other_keys``, it becomes possible to more accurately type the update method:
 
+* Declare a new read-only TypedDict type that does not allow other keys
+* Copy all non-read-only entries to it
+* Make all entries read-only and non-required
+* Union this with an iterable of matching key-value pairs
 
-Backwards Compatibility
-=======================
+For instance::
 
-This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
-inherit from ``TypedMapping`` protocols in a way that changes the resulting
-overloads), so code that inspects ``TypedDict`` types will have to change. This
-is expected to mainly affect type-checkers.
+    class Example(TypedDict):
+        a: int
+        b: NonRequired[str]
+        c: ReadOnly[int]
 
-The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
-enabling its use in older versions of Python.
+    class ExampleUpdateDict(TypedDict, total=False, readonly=True, other_keys=False):
+        a: int
+        b: str
+        # c is not present as it is read-only in Example
 
+    ExampleUpdateEntry = tuple[Literal["a"], int] | tuple[Literal["b"], str]
+    ExampleUpdate = ExampleUpdateDict | Iterable[ExampleUpdateEntry]
 
-Security Implications
+Type checkers should permit any type compatible with this TypedDict to be passed into the update operation. As with :pep:`589`, they may choose to continue permitting TypedDict types that allow other keys as well, to avoid generating false positives.
+
+Backwards compatibility
+=======================
+
+This PEP changes the rules for how TypedDict behaves, so code that inspects TypedDict types will have to change to support types using the new features. This is expected to mainly affect type-checkers.
```

Comment:
```suggestion
This PEP adds new features to ``TypedDict``, so code that inspects ``TypedDict`` types will have to change to support types using the new features. This is expected to mainly affect type-checkers.
```

The existing wording feels too strong; it suggests that existing TypedDict code might change in behavior, when really it's only code that uses the new flags that can change.